### PR TITLE
Add tests for issue-290

### LINKED ISSUE #335: Add tests for issue-290
[PR-328](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/pull/328) is only for code fix of [issue-290](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/issues/290). It would nice to have tests for this.

### PR DESCRIPTION
**Pull request by huchen2021**
Fixes #335

Signed-off-by: chen hui <huchen@vmware.com>

### CODE REVIEW

File: controllers/infrastructure/byomachine_controller.go [hunk 1]
```diff
@@ -259,27 +259,25 @@ func (r *ByoMachineReconciler) reconcileNormal(ctx context.Context, machineScope
...
-		}
-
-		providerID, err := r.setNodeProviderID(ctx, remoteClient, machineScope.ByoHost)
-		if err != nil {
-			logger.Error(err, "failed to set node providerID")
-			r.Recorder.Eventf(machineScope.ByoMachine, corev1.EventTypeWarning, "SetNodeProviderFailed", "Node %s does not exist", machineScope.ByoHost.Name)
-			return ctrl.Result{}, err
-		}
+	logger.Info("Updating Node with ProviderID")
+	remoteClient, err := r.getRemoteClient(ctx, machineScope.ByoMachine)
```
Comment (anusha94) [inline controllers/infrastructure/byomachine_controller.go:263]: @huchen2021 
I haven't completely gone through the PR, but I assumed this PR was only to add tests. Why the changes in `byomachine_controller`?

Comment (huchen2021) [reply controllers/infrastructure/byomachine_controller.go:263]: There are two code changes as followed:
1) One is for removing the if line, it introduced by @dharmjit. The original behavior: " If machineScope.ByoMachine.Spec.ProviderID is empty, don't call setNodeProviderID function". This breaks code of issue-290. Even machineScope.ByoMachine.Spec.ProviderID is empty, it may has invalid value, and it shouldn't just let it go. In the code of issue-290, inside of function setNodeProviderID, it will check the value of machineScope.ByoMachine.Spec.ProviderID and do according process.

2) The other is to Capitalize the first letter of providerIDSuffixLength to export it, then controll_test.go can use the variable.

Comment (dharmjit) [reply controllers/infrastructure/byomachine_controller.go:263]: > dharmjit and I almost check in the code in the same time

@huchen2021, I am not sure which 2 changes were pushed at the same time. The `if` check is not introduced lately.

Comment (huchen2021) [reply controllers/infrastructure/byomachine_controller.go:263]: > > dharmjit and I almost check in the code in the same time
> 
> @huchen2021, I am not sure which 2 changes were pushed at the same time. The `if` check is not introduced lately.

Sorry, i read it wrong from "5 months ago" to "5 days ago".
Anyway, this if check should be removed.
Sorry for that again.

File: controllers/infrastructure/byomachine_controller_test.go [hunk 2]
```diff
@@ -117,6 +118,32 @@ var _ = Describe("Controllers/ByomachineController", func() {
...
+			BeforeEach(func() {
+				byoHost = builder.ByoHost(defaultNamespace, "test-node-providerid-host").Build()
+				Expect(k8sClientUncached.Create(ctx, byoHost)).Should(Succeed())
+			})
+
+			It("should not return error when node.Spec.ProviderID is with correct value", func() {
+				node := builder.Node(defaultNamespace, byoHost.Name).Build()
+				node.Spec.ProviderID = fmt.Sprintf("%s%s/%s", controllers.ProviderIDPrefix, byoHost.Name, util.RandomString(controllers.ProviderIDSuffixLength))
+				Expect(clientFake.Create(ctx, node)).Should(Succeed())
+				WaitForObjectsToBePopulatedInCache(byoHost)
```
Comment (anusha94) [inline controllers/infrastructure/byomachine_controller_test.go:132]: Can we move `WaitForObjectsToBePopulatedInCache` call to `BeforeEach` where byoHost is actually created?

File: controllers/infrastructure/byomachine_controller_test.go [hunk 3]
```diff
@@ -117,6 +118,32 @@ var _ = Describe("Controllers/ByomachineController", func() {
...
+		Context("When node.Spec.ProviderID is already set", func() {
+
+			BeforeEach(func() {
+				byoHost = builder.ByoHost(defaultNamespace, "test-node-providerid-host").Build()
+				Expect(k8sClientUncached.Create(ctx, byoHost)).Should(Succeed())
+			})
+
+			It("should not return error when node.Spec.ProviderID is with correct value", func() {
+				node := builder.Node(defaultNamespace, byoHost.Name).Build()
+				node.Spec.ProviderID = fmt.Sprintf("%s%s/%s", controllers.ProviderIDPrefix, byoHost.Name, util.RandomString(controllers.ProviderIDSuffixLength))
```
Comment (anusha94) [inline controllers/infrastructure/byomachine_controller_test.go:130]: I think it will be useful to add a method in the builder package to set provider id. Code will be more cleaner.

File: controllers/infrastructure/byomachine_controller_test.go [hunk 4]
```diff
@@ -117,6 +118,32 @@ var _ = Describe("Controllers/ByomachineController", func() {
...
+				Expect(clientFake.Create(ctx, node)).Should(Succeed())
+				WaitForObjectsToBePopulatedInCache(byoHost)
+				_, err := reconciler.Reconcile(ctx, reconcile.Request{NamespacedName: byoMachineLookupKey})
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("should return error when node.Spec.ProviderID is without correct value", func() {
```
Comment (anusha94) [inline controllers/infrastructure/byomachine_controller_test.go:137]:
```suggestion
			It("should return error when node.Spec.ProviderID has stale value", func() {
```

File: controllers/infrastructure/byomachine_controller_test.go (hunk 3 above)
Comment (huchen2021) [reply controllers/infrastructure/byomachine_controller_test.go]: ok.

File: controllers/infrastructure/byomachine_controller_test.go (hunk 2 above)
Comment (huchen2021) [reply controllers/infrastructure/byomachine_controller_test.go]: ok.

**Review by dharmjit — DISMISSED** [review]
lgtm

File: test/builder/builders.go [hunk 5]
```diff
@@ -355,7 +362,9 @@ func (n *NodeBuilder) Build() *corev1.Node {
 			Name:      n.name,
 			Namespace: n.namespace,
 		},
-		Spec:   corev1.NodeSpec{},
+		Spec: corev1.NodeSpec{
```
Comment (anusha94) [inline test/builder/builders.go:365]: Wonder if we should set this only if `providerID` is present. Like this -
https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/blob/9663cb8e24146dcec338c95228efff61f291f5c9/test/builder/builders.go#L76-L78

File: test/builder/builders.go [hunk 6]
```diff
@@ -344,6 +345,12 @@ func Node(namespace, name string) *NodeBuilder {
 	}
 }
 
+// WithProviderID adds the passed providerID to the ByoHostBuilder
```
Comment (anusha94) [inline test/builder/builders.go:348]:
```suggestion
// WithProviderID adds the passed providerID to the NodeBuilder
```

File: test/builder/builders.go (hunk 5 above)
Comment (huchen2021) [reply test/builder/builders.go:365]: No, Selector is pointer, providerID is string. Those value are all present, just whether initialize.
If a pointer is not initialize, it is nil, and it will coredump when you write data to nil address.
If a string is not initialize, it is empty.  It's no difference.

**Review by anusha94 — APPROVED** [review]
lgtm.

**Review by dharmjit — APPROVED** [review]
